### ISSUE TITLE
Missing support for unparenthesized expression list as match subject

### ISSUE THREAD

**Issue by pewscorner** [microsoft/pylance-release#1044]
## Environment data

-   Language Server version: 2021.3.1
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.0a6

## Expected behaviour

According to the syntax section of PEP 634 the new `match` statement takes a subject_expr which may be an unparenthesized list of comma-separated star_named_expressions, and this works fine in Python 3.10.0a6, so Pylance should not complain about it. I realize that Python 3.10.0 hasn't been released yet, but since Pylance seems to already handle most `match` syntax correctly, I thought I'd mention this deficit.

## Actual behaviour

Pylance covers the code in squiggly red lines if the subject_expr contains commas and is not parenthesized.

## Logs

```
User belongs to experiment group 'pythonaa'
User belongs to experiment group 'pythonSendEntireLineToREPL'
User belongs to experiment group 'pythonNotDisplayLinterPrompt'
User belongs to experiment group 'pythonTensorboardExperiment'
User belongs to experiment group 'CollectLSRequestTiming - control'
User belongs to experiment group 'DeprecatePythonPath - experiment'
> conda --version
> pyenv root
> python3.7 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python3.6 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python3 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python2 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3.6 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3.7 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -2 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> conda info --json
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
Starting Pylance language server.
Python interpreter path: D:\Python\Python310\python.exe
Error 2021-03-13 11:44:57: Detection of Python Interpreter for Command python3 and args  failed as file  does not exist
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'D:\Python\Python24\python.exe' [Error: Command failed: "D:\\Python\Python24\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Traceback (most recent call last):
  File "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py", line 9, in ?
    import runpy
ImportError: No module named runpy

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"D:\\\\Python\\Python24\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'D:\Python\Python26\python.exe' [Error: Command failed: "D:\\Python\Python26\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Traceback (most recent call last):
  File "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py", line 30, in <module>
    runpy.run_path(module, run_name="__main__")
AttributeError: 'module' object has no attribute 'run_path'

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"D:\\\\Python\\Python26\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'D:\Python\Python25\python.exe' [Error: Command failed: "D:\\Python\Python25\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Traceback (most recent call last):
  File "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py", line 30, in <module>
    runpy.run_path(module, run_name="__main__")
AttributeError: 'module' object has no attribute 'run_path'

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"D:\\\\Python\\Python25\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-03-13 11:44:59: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.636928669\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.636928669\\pythonFiles\\interpreterInfo.py"'
}
> conda --version
> D:\Python\Python310\python.exe -c "
import unittest
loader = unittest.TestLoader()
suites = loader.discover(".", pattern="*unittest.py")
print("start") #Don't remove this line
for suite in suites._tests:
    for cls in suite._tests:
        try:
            for m in cls._tests:
                print(m.id())
        except:
            pass"
cwd: d:\Documents\Python
> D:\Python\Python310\python.exe -c "
import unittest
loader = unittest.TestLoader()
suites = loader.discover(".", pattern="*unittest.py")
print("start") #Don't remove this line
for suite in suites._tests:
    for cls in suite._tests:
        try:
            for m in cls._tests:
                print(m.id())
        except:
            pass"
cwd: d:\Documents\Python
```

### Without parentheses (bad)

```python
match 1, 2:
    case x, y:
        print(x, y)
```

![image](https://user-images.githubusercontent.com/15351477/111027738-77ca4100-83f2-11eb-89b7-6ffa4d7dafc0.png)

### With parentheses (OK)

```python
match (1, 2):
    case x, y:
        print(x, y)
```

![image](https://user-images.githubusercontent.com/15351477/111027779-b3650b00-83f2-11eb-8e53-a99d90dc0d6f.png)

**Comment by pewscorner** [comment]
Note that even the syntax highlighting here on Github doesn't recognize the `match` keyword in the code snippet with the unparenthesized subject.

**Comment by erictraut** [comment]
Thanks for pointing that out. I missed that part in PEP 634. I've modified the parser in pylance to support this. This will be fixed in the next release of pylance.

The new pattern-matching feature introduces a significant addition to the Python grammar, so it's going to take a while for tools and libraries to add support. The library that github uses for syntax highlighting is one of many that will need to be updated.

**Comment by pewscorner** [comment]
Sounds great, thanks!

No, I wasn't really expecting any tools to support the new feature yet, but for a moment I thought github was recognizing the `match` keyword in one of my snippets, so that's why I mentioned it. Now I realize that it was simply being colored as a function name due to the parentheses - not as a keyword.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.3.2, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202132-17-march-2021

**Comment by pewscorner** [comment]
It works! Thanks!
